database-commons/src/main/java/io/cdap/plugin/db/source/AbstractDBSource.java

```diff
@@ -191,41 +193,52 @@ private Schema loadSchemaFromDB(Class<? extends Driver> driverClass)
-      String externalDocumentationLink = getExternalDocumentationLink();
-      if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
-        if (!errorMessage.endsWith(".")) {
-          errorMessage = errorMessage + ".";
-        }
-        errorMessage = String.format("%s For more details, see %s", errorMessage, externalDocumentationLink);
-      }
-      throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
-        errorMessage, errorMessageWithDetails, ErrorType.USER, false, ErrorCodeType.SQLSTATE,
-        e.getSQLState(), externalDocumentationLink, new SQLException(e.getMessage(),
-          e.getSQLState(), e.getErrorCode()));
-    } finally {
-      driverCleanup.destroy();
+      return Failsafe.with(RetryPolicyUtil.createConnectionRetryPolicy(sourceConfig.getInitialRetryDuration(),
+        sourceConfig.getMaxRetryDuration(), sourceConfig.getMaxRetryCount()))
+              .get(() -> {
+                try (Connection connection = DriverManager.getConnection(connectionString, connectionProperties)) {
+                  executeInitQueries(connection, sourceConfig.getInitQueries());
+                  return loadSchemaFromDB(connection, sourceConfig.getImportQuery());
+                }
+                catch (SQLException e) {
+                  // wrap exception to ensure SQLException-child instances not exposed to contexts without jdbc
+                  // driver in classpath
+                  String errorMessage =
+                          String.format("SQL Exception occurred: [Message='%s', SQLState='%s', ErrorCode='%s'].",
+                                  e.getMessage(),
+                                  e.getSQLState(), e.getErrorCode());
+                  String errorMessageWithDetails = String.format("Error occurred while trying to" +
+                                  " get schema from database." +
+                          "Error message: '%s'. Error code: '%s'. SQLState: '%s'", e.getMessage(), e.getErrorCode(),
+                          e.getSQLState());
+                  String externalDocumentationLink = getExternalDocumentationLink();
+                  if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
+                    if (!errorMessage.endsWith(".")) {
+                      errorMessage = errorMessage + ".";
+                    }
+                    errorMessage = String.format("%s For more details, see %s", errorMessage,
+                            externalDocumentationLink);
+                  }
+                  throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+                          errorMessage, errorMessageWithDetails, ErrorType.USER, false, ErrorCodeType.SQLSTATE,
+                          e.getSQLState(), externalDocumentationLink, new SQLException(e.getMessage(),
+                                  e.getSQLState(), e.getErrorCode()));
```

Comment:
If `loadSchemaFromDB(Connection connection, String query)` already includes retry logic, we don't need to add another layer here, as having nested retries can lead to undesirable behavior.

Also `Failsafe` will wrap any non-retryable exception or an exception after all retries are exhausted in a `FailsafeException` so where ever we are doing `catch(SQLException)` it won't work, you might need to catch `FailsafeException` & check cause.